experiments/xls/stream_logging.py

```diff
@@ -21,4 +23,44 @@ def __exit__(self, exc_type, exc_value, traceback):
         xls_logger.removeHandler(self.handler)
 
     def get_log(self) -> str:
-        return self._stream.getvalue()
\ No newline at end of file
+        return self._stream.getvalue()
+
+
+class LogParser(object):
+    """
+    Current logging records are generated as strings with the following
+    format "LEVEL::MESSAGE". This parser takes the raw log strings and turns
+    them into a dictionary for ease of access of the level and message props.
+    """
+
+    def __init__(self, raw_logs: [str] = []):
+        self.raw_logs = raw_logs
+        self.logs = []
+        self.parse_logs(self.raw_logs)
+
+    def parse_logs(self, raw_logs: [str]):
+        for log in raw_logs:
+            self.parse_log_line(log)
+
+    def parse_log_line(self, raw_log: str):
+        """
+        Example:
+            - input [raw_log] = "INFO::Imported with id 99"
+            - output [dict] = {"level": "INFO", "message": "Imported with id 99"}
+        """
+        if not raw_log:
+            return
+        log_tuple = raw_log.split('::')
+        if len(log_tuple) == 2:
+            self.add_message(log_tuple[0], log_tuple[1])
+        else:
+            self.add_error_message(raw_log)
+
+    def add_message(self, level: str, message: str) -> dict():
+        self.logs.append({"level": level, "message": message})
+
+    def add_error_message(self, message: str) -> dict():
+        self.add("ERROR", message)
```

Comment:
since you renamed the `add` method, it must be `add_message`

Reply:
missed that